### PR TITLE
refactor(nat): refactor gateway resource

### PR DESCRIPTION
**Pull request by Lance52259**
**What this PR does / why we need it**:
Rewrite the NAT gateway resource to replace the auto-generated codes.

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. rewrite the NAT gateway resource.
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/nat' TESTARGS='-run=TestAccGateway_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/nat -v -run=TestAccGateway_basic -timeout 360m -parallel 4
=== RUN   TestAccGateway_basic
=== PAUSE TestAccGateway_basic
=== CONT  TestAccGateway_basic
--- PASS: TestAccGateway_basic (96.32s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/nat       96.408s
```

### CODE REVIEW

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+func resourceGatewayUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.NatGatewayClient(region)
+	if err != nil {
+		return diag.Errorf("error creating NAT v2 client: %s", err)
+	}
+
+	gatewayId := d.Id()
```
Comment (hmbig2) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: add haschange("name",spec,desc)

File: docs/resources/nat_gateway.md [hunk 2]
```diff
@@ -4,49 +4,52 @@ subcategory: "NAT Gateway (NAT)"
 
 # huaweicloud_nat_gateway
 
-Manages a Nat gateway resource within HuaweiCloud Nat.
+Manages a Nat gateway resource within HuaweiCloud.
```
Comment (ShiChangkuo) [inline docs/resources/nat_gateway.md]:
```suggestion
Manages a **public** NAT gateway resource within HuaweiCloud.
```

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+		return diag.Errorf("error creating NAT v2 client: %s", err)
+	}
+
+	opts := gateways.CreateOpts{
+		Name:                d.Get("name").(string),
+		VpcId:               d.Get("vpc_id").(string),
+		InternalNetworkId:   d.Get("subnet_id").(string),
+		Spec:                d.Get("spec").(string),
+		Description:         d.Get("description").(string),
+		EnterpriseProjectId: common.GetEnterpriseProjectID(d, cfg),
```
Comment (ShiChangkuo) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]:
```suggestion
		EnterpriseProjectId: cfg.GetEnterpriseProjectID(d),
```

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+					string(PublicSpecTypeSmall),
+					string(PublicSpecTypeMedium),
+					string(PublicSpecTypeLarge),
+					string(PublicSpecTypeExtraLarge),
+				}, false),
+				Description: "The specification of the NAT gateway.",
+			},
+			"vpc_id": {
+				Type:        schema.TypeString,
+				Optional:    true,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: it's required in the doc

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+				Computed:    true,
+				ForceNew:    true,
+				Description: "The ID of the VPC to which the NAT gateway belongs.",
+			},
+			"subnet_id": {
+				Type:     schema.TypeString,
+				Optional: true,
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: it's required in the doc

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+		Timeout:      d.Timeout(schema.TimeoutDelete),
+		Delay:        5 * time.Second,
+		PollInterval: 10 * time.Second,
+	}
+	_, err = stateConf.WaitForStateContext(ctx)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	if gatewayTags := d.Get("tags").(map[string]interface{}); len(gatewayTags) > 0 {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: I think using GetOk is a more common way

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+		Description: utils.String(d.Get("description").(string)),
+	}
+	_, err = gateways.Update(client, gatewayId, opts)
+	if err != nil {
+		return diag.Errorf("error updating NAT gateway (%s): %s", gatewayId, err)
+	}
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"PENDING"},
+		Target:       []string{"COMPLETED"},
+		Refresh:      gatewayStateRefreshFunc(client, d.Id(), []string{"ACTIVE"}),
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]:
```suggestion
		Refresh:      gatewayStateRefreshFunc(client, gatewayId, []string{"ACTIVE"}),
```

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+		d.Set("region", region),
+		d.Set("name", resp.Name),
+		d.Set("spec", resp.Spec),
+		d.Set("vpc_id", resp.RouterId),
+		d.Set("subnet_id", resp.InternalNetworkId),
+		d.Set("description", resp.Description),
+		d.Set("enterprise_project_id", resp.EnterpriseProjectId),
+		d.Set("status", resp.Status),
+	)
+	gatewayTags, err := tags.Get(networkClient, "nat_gateways", d.Id()).Extract()
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]:
```suggestion
	gatewayTags, err := tags.Get(networkClient, "nat_gateways", gatewayId).Extract()
```

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+func ResourceGateway() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceGatewayCreate,
+		ReadContext:   resourceGatewayRead,
+		UpdateContext: resourceGatewayUpdate,
+		DeleteContext: resourceGatewayDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: missing timeouts block

File: docs/resources/nat_gateway.md [hunk 10]
```diff
@@ -56,19 +59,19 @@ In addition to all arguments above, the following attributes are exported:
...
+* `status` - The current status of the NAT gateway.
 
 ## Timeouts
 
 This resource provides the following timeouts configuration options:
 
-* `create` - Default is 10 minute.
-* `delete` - Default is 10 minute.
+* `create` - Default is 10 minutes.
+* `delete` - Default is 10 minutes.
```
Comment (Jason-Zhang9309) [inline docs/resources/nat_gateway.md]: missing timeout of update

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 1 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: done.

File: docs/resources/nat_gateway.md (hunk 2 above)
Comment (Lance52259) [reply docs/resources/nat_gateway.md]: done.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 4 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: fixed.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 5 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: fixed.

File: docs/resources/nat_gateway.md (hunk 10 above)
Comment (Lance52259) [reply docs/resources/nat_gateway.md]: fixed.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 9 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: fixed.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 8 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: done.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 7 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: done.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 6 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: done.

File: huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go (hunk 3 above)
Comment (Lance52259) [reply huaweicloud/services/nat/resource_huaweicloud_nat_gateway.go]: done.

**Comment by Lance52259** [comment]
```
make testacc TEST='./huaweicloud/services/acceptance/nat' TESTARGS='-run=TestAccPublicGateway_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/nat -v -run=TestAccPublicGateway_basic -timeout 360m -parallel 4
=== RUN   TestAccPublicGateway_basic
=== PAUSE TestAccPublicGateway_basic
=== CONT  TestAccPublicGateway_basic
--- PASS: TestAccPublicGateway_basic (104.07s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/nat       104.201s
```

**Pull request merged**